### PR TITLE
Remove DolphinWX

### PR DESCRIPTION
**Pull request by spycrab**
Just testing the waters and seeing what issues will occur.

**This obviously should only get merged once most of the Qt regressions and (minor) missing features are implemented**

### CODE REVIEW

**Comment by Helios747** [comment]
I'm fine with this once the more severe issues logged on the tracker are resolved.

**Comment by Zexaron** [comment]
There's some in `Base.props` , and the `WXWOverrides.props` looks less sure.
There is a WX define in GekkoDisassembler.cpp - Line 1019
GLInterfaceBase.h - Line 60
Common.h - Line 65
MsgHandler.cpp - Line 27 and 33
BootManager.h - Line 12
Gekko.h - Line 274 not sure
RenderBase.h - Line 174 not sure

_I searched in-browser and did not found any on the Files Changed page, I scrolled all the page to make sure the page loaded all the content and it seems like it did. However after a while the tab crashed in Opera with an "out of memory" error a few times but it was generally fine when I avoided pressing any buttons, search worked fine._
`VSProps` _folder nor_ `.props` _seem to be ignored, and my VS did not modify these files, based on that I thought it's okay to raise this, I also tried to learn if there's some kind of auto-updating that will remove these lines once the main project is removed but found no info on that, but it probably wouldn't hurt if these were removed right here._

**Comment by spycrab** [comment]
Most of the stuff you mentioned should be removed later on.

> There's some in Base.props , and the WXWOverrides.props looks less sure.

That should probably be removed with this PR though.

**Comment by lioncash** [comment]
> `GekkoDisassembler.cpp` - Line 1019
> `Gekko.h` - Line 274

These two are unrelated to wxWidgets

**Comment by Ebola16** [comment]
I haven't been able to successfully compile QT2 on Win 10 in Visual Studio using the instructions [here](https://github.com/dolphin-emu/dolphin/wiki/Building-With-Qt) . I keep getting the attached error. I've also made a [Dolphin forum post](https://forums.dolphin-emu.org/Thread-compiling-dolphinqt2). This needs to be resolved before getting rid of DolphinWX.
capture

**Comment by fancymelonub** [comment]
Make sure to add "Path to Virtual SD Card" from WX to QT before merging

**Comment by spycrab** [comment]
> Make sure to add "Path to Virtual SD Card" from WX to QT before merging

That was added over two weeks ago...

**Comment by sepalani** [comment]
Regardless, IMHO, I don't think this PR should be blindly merged.

ATM, there are too many annoyances/flaws that prevent some implemented features to be used at their fullest on Qt. That is/was true, especially for the debugger part of Dolphin. Even if WxWidgets works but Qt has nice additions over it, it shouldn't be used as an excuse to merge this PR too early. I think that's also why @spycrab said that it won't be merged before Qt regressions are fixed and missing features implemented.

Dolphin still has parts where WxWidgets is uglier but works better than its Qt counterpart while code-wise it's probably cleaner on Qt but still needs to be polished.

**Comment by spycrab** [comment]
We're reaching a state where I'd be more confident in this being mergable.
Any opinions? Especially if there are certain things we should wait for?

cc: @JMC47 @MayImilae

**Comment by spycrab** [comment]
> Dolphin.exe is being compiled in VS but the debugger is using DolphinWx.exe by default.

This PR makes it default to Qt now but keep in mind that your user preferences might still override this.

**Comment by Zexaron** [comment]
My remaining tickets are few and low priority, I have nothing pending, seen nothing suspicious since.  Reminder about `base.props` and `WXWOverrides.props`

**Comment by MayImilae** [comment]
My opinion is that we should not do remove WX yet. Whether technically ok now or not, it's just too soon. We need to give several months of just leaving things be and letting users get used to WX before we remove it.

So, I do not agree with merging this before 2019.

**Comment by Ebola16** [comment]
> This PR makes it default to Qt now but keep in mind that your user preferences might still override this.

This PR correctly defaults to Dolphin.exe (instead of DolphinWx.exe) in VS debugger. Any chance you could create a separate PR for that fix so it can get merged sooner?

**Comment by spycrab** [comment]
> Even when all issues are fixed, there's no rush to remove it.

Sure.

To clarifiy: I have no problem with waiting a few more months but

> I absolutely do not agree with merging this before 2019.

Is just over the top. I think September of this year is a more realistic date that should allow for enough time to have everything sorted out more than thoroughly.

Also keep in mind that many changes in Core require updating wx as well and we all know what that entails...

**Comment by Zexaron** [comment]
Just a thought: What about branching Qt improvement to a Wx-less one while the Qt momentum is still high and memory is fresh, I can only speak for myself but If I had a big pause in between I would take some time to get back into gears. And the stuff I mentioned in my GUI [thread](https://forums.dolphin-emu.org/Thread-gui-graphics-window-renaming-and-sort-hacks-advanced-and-debug-tabs?page=4) would be way easier without Wx, but none of that is a strong opinion, so no need to take it that seriously.

**Comment by Helios747** [comment]
... Why

you gain nothing.

**Comment by Zexaron** [comment]
I don't know if it's worth it or not, I was merely suggesting that to those who were planning it. AFAIK it was decided way long ago that first Qt would copy Wx design, only then changes would come and would only be on Qt from that point forward. I stopped bothering after that decision was made, except the stuff I had stashed prior to that, that's what I posted in the thread for anyone to take into account if they wish, that's it from me.

**Comment by Helios747** [comment]
We don't need to branch with Wx entirely removed.......

We can just.... Ignore that Wx is there.

**Comment by MayImilae** [comment]
> Is just over the top. I think September of this year is a more realistic date that should allow for enough time to have everything sorted out more than thoroughly.

My concern isn't about your ability to fix all of the Qt issues. I know that's going to happen pretty soon! What I'm worried about is how users react to the full removal of WX. We want them to like Qt! But a lot of users may have tried Qt immediately and switched back to avoid bugs, or just refused to change over outright. If we just yank WX away not even 6 months after Qt was made default, we're going to have protests. But if we let WX rot for a year and then remove it, it makes it a *much* easier sell to users on the change, and reduces the amount of holdovers that will be abruptly affected by it.

No one here likes WX, but, a lot of users do, and it's the UI we've had for literally a decade! We just need to give users _time_ to acclimate to the change and migrate over.

Besides, it doesn't really hurt us to let WX rot a bit, but it could certainly help us sell Qt.

**Comment by spycrab** [comment]
> Besides, it doesn't really hurt us to let WX rot a bit, but it could really help us sell Qt.

Except when it does though. Just today I had to introduce a rather nasty hack into Qt because of Wx.

And there's also this:

> Also keep in mind that many changes in Core require updating wx as well and we all know what that entails...

Keeping Wx around definitely has a cost.

**Comment by spycrab** [comment]
Just to be clear: I don't have any problems waiting a bit longer for users to get used to Qt and convincing as many as possible to switch voluntarily. But six months seem unreasonably long for that.

**Comment by MayImilae** [comment]
Well ok. But my concern still stands though. If users hate us for removing WX, that's an incredible failure on our part, and will significantly hurt Qt. We need to be careful, even more careful than we were when we made Qt the default.

Do we have analytics for WX vs Qt users? I have a sinking feeling those numbers are going to be nasty, considering how many users I run into running random old builds for "speed".

**Comment by sepalani** [comment]
@spycrab Truth be told, I don't think Wx is the true problem but rather the abstraction between the core and the UI. Most code from Wx is almost copied/pasted from/to other backends, making changes to one of them a pain since it has to be back-ported to the others most of the time.

The real solution isn't to get rid of Wx since the problem will raise again if someone want to implement another one. I don't know... maybe UWP. So the real solution would rather be to provide a better abstraction between the core and the UI, thus, it will get rid of these copy-paste shenanigans as well.

That's why I agree on @MayImilae point and it's probably a wise decision to let users be used to Qt, rather than forcing them.

**Comment by leoetlino** [comment]
For one thing, DolphinWX is holding back the migration to the new config system. I'm definitely not interested in writing tons of DolphinWX code that will just be thrown away.

**Comment by spycrab** [comment]
Again: I don't have any issues with giving it a few more months. But 2019 is just too much. 

We're already at 60-something percent of users using Qt anyway and that number will only go up as the months pass.

> Truth be told, I don't think Wx is the true problem but rather the abstraction between the core and the UI

Wx carries multiple problems specific to it that don't affect NoGUI or Android. I'm not going to elaborate on it though as I think this is quite self evident (just look at the wx code if you must).

**Comment by Helios747** [comment]
Also agree with "too early now, but fuck waiting until next year"

**Comment by tony971** [comment]
End user here. Couldn't care less about the UI framework as long as it looks/works as expected. I don't think any end user has a special place in their heart for Qt, wx, or any other framework. This is like changing pencil brands to an end user.

Personally, I've looked at this transition from the following viewpoint: "Oh, changing the underlying code helps the devs out? That's cool, I guess. When it hits functional parity, kill the old one with fire. My experience doesn't change in the slightest."

**Comment by Zexaron** [comment]
@tony971 @nbohr1more Sure ... but just a note that Qt used to have vertical tab placement, these points probably wouldn't have the equal amount of strength if it wasn't switched back to horizontal not a long time ago. #6659 (took some time to find it, almost gave up)

**Comment by JosJuice** [comment]
@Zexaron Yes, but now they are placed horizontally, so I don't see why this is relevant.

**Comment by Ebola16** [comment]
When we remove WX, will the command line argument -d still be needed for anything? If not it should also be removed.

**Comment by leoetlino** [comment]
It is still used to activate debug mode. The only difference in DolphinQt2 is that you can also toggle debug mode in the UI.

**Comment by Ebola16** [comment]
Yeah, I guess there's no point in removing that. Nevermind.

**Comment by spycrab** [comment]
Now that we have some improved statistics - some observations:

The whole 'voluntarily' thing doesn't really seem to work. The amount of users using wx is - and always has been since the default switch - stuck at 20%. There is nothing that indicates this will change at any point in the future.

*So no matter when* we switch there'll be 1/5 of all users that could potentially be upset about it.

cc: @MayImilae @JMC47

**Comment by MayImilae** [comment]
In all of our big removals, we've always given at least 6 months warning, but typically it's more like a year to 18 months. We've even waited 2 years in some cases. I know you are really itching to remove WX at this point, but we really just need to give users time.

Plus with more time we can hammer home that we're removing it over multiple Progress Reports to get the userbase as ready for this as possible, both informing any hold outs and making sure people know that WX really is going away and they won't be surprised. And hopefully that will reduce that 20% figure even more, but if not, it will minimize it as much as we can for a reduced (not 18 months) time frame.

**Comment by shuffle2** [comment]
Not sure what repeatedly publicizing that Dolphin's going to switch UI frameworks accomplishes (especially since it's already been done).  
If Qt has feature parity and the new "dolphin.exe" is Qt instead of WX, what does it matter when exactly the change happens? Dolphin hasn't even had a stable release for ages at this point.
Just Do It.

**Comment by Dockolm** [comment]
Then we should at least wait the next progress report and clearly announcing WX will be dropped sooner than expected, probably before the next PR.
Qt's UI being very close to WX, it shouldn't be such a big deal.

**Comment by orbea** [comment]
I agree with @MayImilae. Deprecated features should be given time, I personally would suggest making at least one more stable release with wx.

**Comment by tony971** [comment]
To be fair, this isn't your standard depreciated feature. As Qt hits feature parity, there won't be a use case in which Wx is more useful. Compare that to say, DX9, and you could easily name some instances in which some users would be impacted by its removal.

**Comment by MayImilae** [comment]
@orbea We have already stated that WX will be removed before the next stable release.

**Comment by MayImilae** [comment]
That's a great idea! If we add some some of WX warning about it being deprecated (akin to the Qt warning) and then wait a few months, that would be a great approach.

**Comment by spycrab** [comment]
Wx being around definitely hinders implementing new features in Qt. If we want to keep around longer it will obviously negatively impact its progress. Hope that's clear.

**Comment by MayImilae** [comment]
We should not just pull the rug out from thousands of users. :/ Yes it may suck to keep WX around for several more months, but we have to do this properly and with great care. We want everyone to celebrate the removal of WX, and not despise it and make a bloody WX fork. And yes that could happen; we've seen it happen for far sillier things.

**Comment by spycrab** [comment]
> Yes it may suck to keep WX around for several more months

Several more months without new Qt features it is then.

**Comment by Bombastisch** [comment]
I am sorry to pop in, but I think I know the reason why there are these constant 20%.
These useres probably have a desktop icon linking to Wx since qt was still a bit buggy in the beginning. If the users then update manually the desktop icon still links to DolphinWX and some of the users might not even notice.

**Comment by JMC47** [comment]
A lot of hte config stuff we wanna do requires significant rewriting, and doing the work on both Wx and Qt jsut isn't reasonable when we're planning on dropping Qt.  I've also noticed that a couple of developers are waiting for Wx to drop before working on features that interact with the GUI.

If people are using old builds for Wx, I don't see why keeping Wx around affects that.  It's not like old savestates and whatnot are going to magically start working in new builds.

**Comment by delroth** [comment]
20% of people *on recent builds* are using Wx. Let's not dodge the facts by
saying that e.g. they use old builds.

**Comment by spycrab** [comment]
Do what you want. I don't care anymore. ``¯\_(ツ)_/¯``

**Comment by newblar** [comment]
I have noticed that on Windows 10 using that search on the taskbar if you type in Dolp best match shows Dolphin.exe but if you type Dolph Windows then shows DolphinWX as best match. Could be why some people keep using WX.

**Comment by Leon422** [comment]
Oh, I wasn't expecting you to do this so soon. I just want to say I'm really happy you did! Moving forward is always a good thing.

File: CMakeLists.txt [hunk 1]
```diff
@@ -695,35 +692,11 @@ if(NOT ANDROID)
...
-    add_subdirectory(Externals/wxWidgets3)
-    set(wxWidgets_FOUND TRUE)
-  endif()
-endif()
 
 if(USE_DISCORD_PRESENCE)
   message(STATUS "Using static DiscordRPC from Externals")
   add_subdirectory(Externals/discord-rpc)
-endif()
+endif()
```
Comment (KAMiKAZOW) [inline CMakeLists.txt]: Why?

**Pull request merged**